apps/accelerate/forward_forward/forward_forward/operations/trainers.py

```diff
@@ -183,3 +188,49 @@ def _train(
             predictions, _ = model.positive_eval(test_data, theta)
             perplexity = compute_perplexity(predictions)
             self.logger.info(f"Perplexity: {perplexity}")
+
+            
+class CNNForwardForwardTrainer(BaseForwardForwardTrainer):
+    def _train(self, epochs: int, theta: float, device: str, **kwargs):
+        model = self.model.to(device)
+        for epoch in range(epochs):
+            accumulated_goodness = None
+            model.train()
+            # Preprocess the data
+            x_train = [data.flatten().numpy() for data, label in self.train_data]
```

Comment:
why are you flattening the input tensors?

Reply:
I do not understand this. So, I get that I should not flatten x_train. Since, x_train is used in training so this means its better to have x_test as unflattened tensor too, right?

Reply:
Here the problem is that the network expects as input not-flattened data, while KMeans expects as input flattened data. Furthermore, we should considered that in forward-forward we should add the negative-examples too for training the model. 

I propose to modify the code in the following way: 
* Instead of creating a new dataloader using the `x_train` numpy data, we use KMeans for obtaining the "unsupervised labels" in a new function `get_unsupervised_label` and then we use the obtained labels to get both the positive and negative data. In the example below the function directly returns the dataloader we should use in the training loop (and the kmeans model).
```python
def get_unsupervised_label(self, device):
    x_train = np.concatenate(
        [data.detach().cpu().numpy() for data, label in
         self.train_data], axis=0)

    # Perform clustering
    kmeans = KMeans(n_clusters=10, random_state=0)
    train_labels = kmeans.fit_predict(x_train)
    train_labels = torch.from_numpy(train_labels).to(device)
    label_injector = LabelsInjector([f"cluster_{i}" for i in range(10)])

    train_bs = self.train_data.batch_size
    progressive_train_dataset = ProgressiveTrainingDataset(
        (label_injector.inject_train(
            x.unsqueeze(1), 
            train_labels[i*train_bs: (i+1)*train_bs]
        ) for i, (x, y) in enumerate(self.train_data))
    )
    progressive_train_dataloader = torch.utils.data.DataLoader(
        progressive_train_dataset,
        batch_size=2 * train_bs, shuffle=False
    )
    return progressive_train_dataloader, kmeans
```
* When computing the validation metrics we can compute the kmeans prediction on the fly and then use the generated label for computing accuracy. I expect the validation loop to be something like
```python
correct = 0
with torch.no_grad():
    for data, target in self.test_data:
        data = data.to(device)
        numpy_data = data.flatten().cpu().numpy()
        target = torch.from_numpy(kmeans.predict(numpy_data)).to(device)
        pred, _ = model.positive_eval(data.unsqueeze(1), theta)
        correct += pred.eq(target.view_as(pred)).sum().item()
self.logger.info(
    f"Test accuracy: {correct} / 10000 ({correct / 10000 * 100}%)"
)
```

What do you think?

Reply:
I sort of understand what you are saying. I did something similar [here](https://github.com/cyclotomicextension/Contrastive-Learning-via-FFA/blob/main/ffa_implemention_of_contrastive_learning.py). I shall work on this now! Thanks a lot